homeassistant/components/icloud/config_flow.py

```diff
@@ -0,0 +1,224 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
+                    ): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        errors = {}
+
+        icloud_dir = self.hass.config.path("icloud")
+        if not os.path.exists(icloud_dir):
+            await self.hass.async_add_executor_job(os.makedirs, icloud_dir)
+
+        if user_input is None:
+            return await self._show_setup_form(user_input, errors)
+
+        self._username = user_input[CONF_USERNAME]
+        self._password = user_input[CONF_PASSWORD]
+        self._account_name = user_input.get(CONF_ACCOUNT_NAME)
+        self._max_interval = user_input.get(CONF_MAX_INTERVAL, DEFAULT_MAX_INTERVAL)
+        self._gps_accuracy_threshold = user_input.get(
+            CONF_GPS_ACCURACY_THRESHOLD, DEFAULT_GPS_ACCURACY_THRESHOLD
+        )
+
+        if self._configuration_exists(self._username, self._account_name):
+            errors[CONF_USERNAME] = "username_exists"
+            return await self._show_setup_form(user_input, errors)
+
+        try:
+            self.api = await self.hass.async_add_executor_job(
+                PyiCloudService, self._username, self._password, icloud_dir
+            )
+        except PyiCloudFailedLoginException as error:
+            _LOGGER.error("Error logging into iCloud service: %s", error)
+            self.api = None
+            errors[CONF_USERNAME] = "login"
+            return await self._show_setup_form(user_input, errors)
+
+        if self.api.requires_2fa:
+            return await self.async_step_trusted_device()
+
+        return self.async_create_entry(
+            title=self._username,
+            data={
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_ACCOUNT_NAME: self._account_name,
+                CONF_MAX_INTERVAL: self._max_interval,
+                CONF_GPS_ACCURACY_THRESHOLD: self._gps_accuracy_threshold,
+            },
+        )
+
+    async def async_step_import(self, user_input):
+        """Import a config entry."""
+        if self._configuration_exists(
+            user_input[CONF_USERNAME], user_input.get(CONF_ACCOUNT_NAME)
+        ):
+            return self.async_abort(reason="username_exists")
+
+        return await self.async_step_user(user_input)
+
+    async def async_step_trusted_device(self, user_input=None, errors=None):
+        """We need a trusted device."""
+        if errors is None:
+            errors = {}
+
+        trusted_devices = {}
+        devices = self.api.trusted_devices
```

Comment:
Is this doing I/O ?

Reply:
Yep but can't use executor here

<img width="1049" alt="Capture d’écran 2019-12-02 à 13 59 30" src="https://user-images.githubusercontent.com/14821482/69961651-93b31780-150c-11ea-9717-fe209f36755c.png">

Reply:
Maybe use a lambda in between? Properties can't be called.

Reply:
Thanks @balloob !